src/azure-cli/azure/cli/command_modules/acs/_roleassignments.py

```diff
@@ -0,0 +1,168 @@
+            raise AzCLIError(err.format(resource_group_name))
+    elif resource_group_name:
+        scope = subscription_scope + '/resourceGroups/' + resource_group_name
+    else:
+        scope = subscription_scope
+    return scope
+
+
+def create_role_assignment(
+    cmd,
+    role,
+    assignee,
+    is_service_principal,
+    resource_group_name=None,
+    scope=None,
+    resolve_assignee=True,
+):
+    return _create_role_assignment(
+        cmd,
+        role,
+        assignee,
+        resource_group_name,
+        scope,
+        resolve_assignee=(is_service_principal and resolve_assignee),
+    )
+
+
+def _create_role_assignment(cmd, role, assignee,
+                            resource_group_name=None, scope=None, resolve_assignee=True):
+    factory = get_auth_management_client(cmd.cli_ctx, scope)
+    assignments_client = factory.role_assignments
+    definitions_client = factory.role_definitions
+
+    # FIXME: is this necessary?
+    if assignments_client.config is None:
+        raise AzCLIError("Assignments client config is undefined.")
+
+    scope = build_role_scope(
+        resource_group_name, scope, assignments_client.config.subscription_id)
+
+    # XXX: if role is uuid, this function's output cannot be used as role assignment defintion id
+    # ref: https://github.com/Azure/azure-cli/issues/2458
+    role_id = resolve_role_id(role, scope, definitions_client)
+
+    # If the cluster has service principal resolve the service principal client id to get the object id,
+    # if not use MSI object id.
+    object_id = resolve_object_id(
+        cmd.cli_ctx, assignee) if resolve_assignee else assignee
+
+    assignment_name = uuid.uuid4()
+    custom_headers = None
+
+    RoleAssignmentCreateParameters = get_sdk(cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION,
+                                             'RoleAssignmentCreateParameters', mod='models',
+                                             operation_group='role_assignments')
+    if cmd.supported_api_version(min_api='2018-01-01-preview', resource_type=ResourceType.MGMT_AUTHORIZATION):
+        parameters = RoleAssignmentCreateParameters(
+            role_definition_id=role_id, principal_id=object_id)
+        return assignments_client.create(scope, assignment_name, parameters, custom_headers=custom_headers)
+
+    # for backward compatibility
+    RoleAssignmentProperties = get_sdk(cmd.cli_ctx, ResourceType.MGMT_AUTHORIZATION,
+                                       'RoleAssignmentProperties', mod='models',
+                                       operation_group='role_assignments')
+    properties = RoleAssignmentProperties(role_definition_id=role_id, principal_id=object_id)
+    return assignments_client.create(scope, assignment_name, properties, custom_headers=custom_headers)
+
+
+def add_role_assignment(cmd, role, service_principal_msi_id, is_service_principal=True, delay=2, scope=None):
```

Comment:
For reference, [this](https://github.com/Azure/azure-cli/blob/dev/src/azure-cli/azure/cli/command_modules/acs/custom.py#L542) is the corresponding implementation in `custom.py`.